l2geth/rollup/fees/rollup_fee.go

```diff
@@ -100,6 +103,63 @@ func DecodeL2GasLimitU64(gasLimit uint64) uint64 {
 	return scaled * tenThousand
 }
 
+type Message interface {
+	From() common.Address
+	To() *common.Address
+	GasPrice() *big.Int
+	Gas() uint64
+	Value() *big.Int
+	Nonce() uint64
+	Data() []byte
+}
+
+type StateDb interface {
+	GetState(common.Address, common.Hash) common.Hash
+}
+
+// TODO: possibly delete this
+func CalculateFee(msg Message, state StateDb, gasUsed *big.Int) (*big.Int, error) {
+	var tx *types.Transaction
+	if msg.To() == nil {
+		tx = types.NewContractCreation(
+			msg.Nonce(),
+			msg.Value(),
+			msg.Gas(),
+			msg.GasPrice(),
+			msg.Data(),
+		)
+	} else {
+		tx = types.NewTransaction(
+			msg.Nonce(),
+			*msg.To(),
+			msg.Value(),
+			msg.Gas(),
+			msg.GasPrice(),
+			msg.Data(),
+		)
+	}
+
+	data := new(bytes.Buffer)
+	if err := tx.EncodeRLP(data); err != nil {
+		return nil, err
+	}
+	// Account for the signature
+	data.Write(bytes.Repeat([]byte("ff"), 68))
+
+	ovalue := state.GetState(rcfg.L2GasPriceOracleAddress, rcfg.OverheadSlot)
+	overhead := ovalue.Big()
+
+	gvalue := state.GetState(rcfg.L2GasPriceOracleAddress, rcfg.L1GasPriceSlot)
+	l1GasPrice := gvalue.Big()
+
+	l1GasUsed := CalculateL1Cost(data.Bytes(), overhead.Uint64())
+	l1Fee := new(big.Int).Mul(l1GasPrice, l1GasUsed)
+
```

Comment:
Note: this scheme changes based on the actual amount of gas used, not the gas limit